libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo3/runtime/java/network/http/internal/BatchingHttpInterceptor.java

```diff
@@ -0,0 +1,256 @@
+import com.apollographql.apollo3.api.json.BufferedSourceJsonReader;
+import com.apollographql.apollo3.exception.ApolloHttpException;
+import com.apollographql.apollo3.exception.ApolloNetworkException;
+import com.apollographql.apollo3.runtime.java.network.http.HttpCallback;
+import com.apollographql.apollo3.runtime.java.network.http.HttpInterceptor;
+import com.apollographql.apollo3.runtime.java.network.http.HttpInterceptorChain;
+import okio.Buffer;
+import okio.BufferedSink;
+import okio.BufferedSource;
+import okio.ByteString;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import static com.apollographql.apollo3.api.Adapters.AnyAdapter;
+
+public class BatchingHttpInterceptor implements HttpInterceptor {
+  private final long batchIntervalMillis;
+  private final int maxBatchSize;
+  private final boolean exposeErrorBody;
+
+  private @NotNull HttpInterceptorChain interceptorChain;
+  private final List<PendingRequest> pendingRequests = new ArrayList<>();
+  private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+  public BatchingHttpInterceptor(long batchIntervalMillis, int maxBatchSize, boolean exposeErrorBody) {
+    this.batchIntervalMillis = batchIntervalMillis;
+    this.maxBatchSize = maxBatchSize;
+    this.exposeErrorBody = exposeErrorBody;
+  }
+
+  private void scheduleExecutePendingRequests() {
+    executor.execute(() -> {
+      try {
+        Thread.sleep(batchIntervalMillis);
+      } catch (InterruptedException ignored) {
+      }
+      executePendingRequests();
+    });
+  }
+
+  private static class PendingRequest {
+    private final HttpRequest request;
+    private final CountDownLatch latch = new CountDownLatch(1);
+    private HttpResponse response;
+    private ApolloNetworkException exception;
+
+    private PendingRequest(HttpRequest request) {
+      this.request = request;
+    }
+
+    private void setResponse(HttpResponse response) {
+      this.response = response;
+      latch.countDown();
+    }
+
+    private void setException(ApolloNetworkException exception) {
+      this.exception = exception;
+      latch.countDown();
+    }
+
+    private HttpResponse getResponse() throws ApolloNetworkException {
+      try {
+        latch.await();
+      } catch (InterruptedException ignored) {
+      }
+      if (exception != null) {
+        throw exception;
+      }
+      return response;
+    }
+  }
+
+  @Override public void intercept(@NotNull HttpRequest request, @NotNull HttpInterceptorChain chain, @NotNull HttpCallback callback) {
+    // Batching is enabled by default, unless explicitly disabled
+    String canBeBatchedHeader = HttpKt.get(request.getHeaders(), ExecutionOptions.CAN_BE_BATCHED);
```

Comment:
Java ergonomics of `Http` aren't great. I added it to https://github.com/apollographql/apollo-kotlin/issues/4171

Reply:
Good catch!